app/src/main/java/org/hwyl/sexytopo/control/graph/GraphView.java

```diff
@@ -1182,40 +1183,57 @@ private void drawSketch(Canvas canvas, Sketch sketch, int alpha) {
             return;
         }
 
+        drawPaint.setAlpha(alpha);
+
+        int lastColour = -1;
         for (PathDetail pathDetail : sketch.getPathDetails()) {
 
             if (!couldBeOnScreen(pathDetail)) {
                 continue;
             }
 
-            drawPaint.setColor(pathDetail.getColour().intValue);
-            drawPaint.setAlpha(alpha);
+            // Avoiding constantly updating the paint colour saves approx. 10% of sketch draw time.
+            // Ideally getPathDetails() would return the paths in colour order but HashSets
+            // are unordered collections
```

Comment:
I suspect that wouldn't save much time in practice - cavers tend to sketch lines in groups of colours anyway (e.g. spend some time sketching black walls, then draw some blue water lines, then draw a different colour of details that are on a different level. Might not be worth the extra complexity.

Reply:
True but the order they sketch them in is irrelevant because they are stored in a `HashSet` so we get them back basically at random

Reply:
Oh I see what you mean. I'd forgotten they were Sets. Maybe they should actually be Lists, and we'd get a minor speed improvement for free. It feels like savefiles should preserve sketching order of details.